pkg/controllers/pod_controller.go

```diff
@@ -0,0 +1,227 @@
+		topologyKey, exists := pod.Annotations[jobset.ExclusiveKey]
+		if !exists {
+			return ctrl.Result{}, nil
+		}
+		topologyValue, err := r.topologyFromPod(ctx, &pod)
+		if err != nil {
+			return ctrl.Result{}, err
+		}
+		if err := r.assignGatedPodsToTopology(ctx, &pod, topologyKey, topologyValue); err != nil {
+			return ctrl.Result{}, err
+		}
+	}
+	return ctrl.Result{}, nil
+}
+
+// SetupWithManager sets up the controller with the Manager.
+func (r *PodReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&corev1.Pod{}).
+		WithEventFilter(predicate.NewPredicateFuncs(func(object client.Object) bool {
+			pod, ok := object.(*corev1.Pod)
+			return ok && usingExclusivePlacement(pod) && isLeaderPod(pod)
+		})).
+		Complete(r)
+}
+
+func SetupPodReconcilerIndexes(ctx context.Context, indexer client.FieldIndexer) error {
+	return indexer.IndexField(ctx, &corev1.Pod{}, podOwnerKey, func(obj client.Object) []string {
+		pod := obj.(*corev1.Pod)
+		owner := metav1.GetControllerOf(pod)
+		if owner == nil {
+			return nil
+		}
+		// Make sure the pod is part of a JobSet using exclusive placement.
+		if _, exists := pod.Annotations[jobset.ExclusiveKey]; !exists {
+			return nil
+		}
+
+		// ...make sure it's a Pod owned by a Job...
+		if owner.Kind != "Job" || owner.APIVersion != batchv1.SchemeGroupVersion.String() {
+			return nil
+		}
+		return []string{owner.Name}
+	})
+}
+
+// nodePoolFromNode gets the node pool name that the given node is part of.
+func (r *PodReconciler) topologyFromPod(ctx context.Context, pod *corev1.Pod) (string, error) {
+	log := ctrl.LoggerFrom(ctx)
+
+	nodeName := pod.Spec.NodeName
+	ns := pod.Namespace
+	topologyKey := pod.Annotations[jobset.ExclusiveKey]
+
+	// Get node the leader pod is running on.
+	var node corev1.Node
+	if err := r.Get(ctx, types.NamespacedName{Name: nodeName, Namespace: ns}, &node); err != nil {
+		// We'll ignore not-found errors, since there is nothing we can do here.
+		// A node may not exist temporarily due to a maintenance event or other scenarios.
+		log.Error(err, fmt.Sprintf("getting node %s", nodeName))
+		return "", client.IgnoreNotFound(err)
+	}
+
+	// Get topology (e.g. node pool name) from node labels.
+	topology, exists := node.Labels[topologyKey]
+	if !exists {
+		return "", fmt.Errorf("node does not have topology label: %s", topology)
+	}
+	return topology, nil
+}
+
+func (r *PodReconciler) assignGatedPodsToTopology(ctx context.Context, leaderPod *corev1.Pod, topologyKey, topologyValue string) error {
+	log := ctrl.LoggerFrom(ctx)
+
+	// Get name of job that owns the leader pod.
+	jobName, err := jobNameFromPod(leaderPod)
+	if err != nil {
+		return err
+	}
+
+	// Get all pods owned by this job.
+	var podList corev1.PodList
+	if err := r.List(ctx, &podList, client.InNamespace(leaderPod.Namespace), client.MatchingFields{podOwnerKey: jobName}); err != nil {
```

Comment:
I don't quite recall, but is the jobName really the ownerKey? did you verify this?

Reply:
I haven't verified, I assumed based on the Job owner key being JobSet name in the jobset controller. I will verify now.

Reply:
Verified the owner name is the job name:

```json
        "ownerReferences": [
            {
                "apiVersion": "batch/v1",
                "blockOwnerDeletion": true,
                "controller": true,
                "kind": "Job",
                "name": "pi",
                "uid": "5a3a1823-c35c-4c30-9002-7bf027638864"
            }
        ],
```

Obtained via running this [example Job](https://kubernetes.io/docs/concepts/workloads/controllers/job/#running-an-example-job) and inspecting the ownerReferences.